File: 12.integer-to-roman/solution.py
class Solution:
    def intToRoman(self, num: int) -> str:
        nums = [num // 1000, num // 100 % 10, num // 10 % 10, num % 10]
        result = duplicate('M', nums[0])
        if nums[1] == 9:
            result += 'CM'
        elif nums[1] >= 5:
            result += 'D'
            result += duplicate('C', nums[1] - 5)
        elif nums[1] == 4:
            result += 'CD'
        else:
            result += duplicate('C', nums[1])
        if nums[2] == 9:
            result += 'XC'
        elif nums[2] >= 5:
            result += 'L'
            result += duplicate('X', nums[2] - 5)
        elif nums[2] == 4:
            result += 'XL'
        else:
            result += duplicate('X', nums[2])
        if nums[3] == 9:
            result += 'IX'
        elif nums[3] >= 5:
            result += 'V'
            result += duplicate('I', nums[3] - 5)
        elif nums[3] == 4:
            result += 'IV'
        else:
            result += duplicate('I', nums[3])
        return result

def duplicate(s: str, i: int) -> str:
    return "" if i <= 0 else duplicate(s, i-1) + s

File: 12.integer-to-roman/solution.go
func intToRoman(num int) string {
	var nums [4]int
	nums[0] = num / 1000
	num %= 1000
	nums[1] = num / 100
	num %= 100
	nums[2] = num / 10
	nums[3] = num % 10
	return iToR(nums[0], "M", "", "") + iToR(nums[1], "C", "D", "M") + iToR(nums[2], "X", "L", "C") + iToR(nums[3], "I", "V", "X")
}

func iToR(num int, one, five, ten string) string {
	switch num {
	case 0:
		return ""
	case 1:
		return one
	case 2:
		return one + one
	case 3:
		return one + one + one
	case 4:
		return one + five
	case 5:
		return five
	case 6:
		return five + one
	case 7:
		return five + one + one
	case 8:
		return five + one + one + one
	case 9:
		return one + ten
	}
	return ""
}
